xpartition.py

```diff
@@ -572,7 +580,8 @@ def write(self, rank):
         region = self.plan.input_partition[rank]
         iData = self.data.isel(region)
         iOut = self.func(iData)
-        iOut.to_zarr(self.path, region=region)
+        unchunked_variables = get_unchunked_variable_names(iOut)
+        iOut.drop_vars(unchunked_variables).to_zarr(self.path, region=region)
```

Comment:
No, it's a fair question.  This was somewhat of a rough edge in the `PartitionMapper` approach even before this change.  For example if `iOut` contains an unchunked data variable things fail in the initialization step with an error like this:
```
>>> import numpy as np; import xarray as xr; import xpartition
>>> func = lambda ds: ds.rename({"a": "renamed"})
>>> data = np.ones((5, 10)); dims = ["x", "y"]; coords = [range(5), range(10)]
>>> a = xr.DataArray(data, dims=dims, coords=coords, name="a")
>>> b = xr.DataArray(data, dims=dims, coords=coords, name="b").chunk({"x": 2})
>>> ds = xr.merge([a, b])
>>> mapper = ds.b.partition.map("store", ranks=3, dims=["x"], func=func, data=ds)
>>> for rank in mapper:
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/spencer/software/xpartition/xpartition.py", line 579, in __iter__
    self._initialize_store()
  File "/Users/spencer/software/xpartition/xpartition.py", line 564, in _initialize_store
    schema = zeros_like(
             ^^^^^^^^^^^
  File "/Users/spencer/software/xpartition/xpartition.py", line 515, in zeros_like
    return ds.apply(
           ^^^^^^^^^
  File "/Users/spencer/software/xarray/xarray/core/dataset.py", line 5993, in apply
    return self.map(func, keep_attrs, args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spencer/software/xarray/xarray/core/dataset.py", line 5964, in map
    variables = {
                ^
  File "/Users/spencer/software/xarray/xarray/core/dataset.py", line 5965, in <dictcomp>
    k: maybe_wrap_array(v, func(v, *args, **kwargs))
                           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spencer/software/xpartition/xpartition.py", line 489, in _zeros_like_dataarray
    chunks = _merge_chunks(arr, override_chunks)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/spencer/software/xpartition/xpartition.py", line 482, in _merge_chunks
    original_chunks = {axis: sizes for axis, sizes in enumerate(arr.chunks)}
                                                      ^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not iterable
```

Technically you previously *could* have gotten away with having an unchunked non-dimension coordinate, but writing it would have been unsafe, since every process would be writing partial data to the same chunk.  For the sake of simplicity I think what I'll do is raise a more informative error in either of these instances.

For context, with the traditional xpartition approach we get around this by writing the unchunked coordinates and data variables fully in the initialization step (and so it is totally valid to drop them when writing the chunked data).  With the `PartitionMapper` approach this gets more complicated, because we initialize the store using a schema that was derived from applying an arbitrary function to a subset of the full dataset, reindexed to the original coordinates.  The subset / reindex complicates things because it means that any unchunked non-dimension coordinates / data variables that either come along for the ride after applying the function or are introduced by the function itself are not necessarily whole when they enter the schema, and so when the schema is initialized it may contain only part of the data for those fields (the rest being filled with NaNs).  The only unchunked fields that are guaranteed to be whole at the schema initialization stage are dimension coordinates (which get written fully then and are of course safe to drop later).  

In a sense having unchunked non-dimension coordinates or data variables is somewhat not well posed with the `PartitionMapper` approach, which in general could probably use some better guardrails / test coverage.

Reply:
Got it, thanks for the detailed explanation! I agree an informative error message is useful here.